cudamapper/src/utils.cpp

```diff
@@ -115,6 +121,190 @@ void print_paf(const std::vector<Overlap>& overlaps,
+{
+    uint32_t* bam_cigar_start = bam_get_cigar(alignment);
+    memcpy(bam_cigar_start, cigar.c_str(), cigar.length());
+    alignment->core.n_cigar = cigar.length();
+
+    return;
+}
+
+static inline void encode_seq(bam1_t* alignment, const std::string& seq)
+{
+    // This table was taken from bam_construct_seq() for base -> nibble encoding
+    static const char L[256] =
+        {
+            15, 15, 15, 15, 15, 15, 15, 15, 15, 15, 15, 15, 15, 15, 15, 15,
+            15, 15, 15, 15, 15, 15, 15, 15, 15, 15, 15, 15, 15, 15, 15, 15,
+            15, 15, 15, 15, 15, 15, 15, 15, 15, 15, 15, 15, 15, 15, 15, 15,
+            15, 15, 15, 15, 15, 15, 15, 15, 15, 15, 15, 15, 15, 0, 15, 15,
+            15, 1, 14, 2, 13, 15, 15, 4, 11, 15, 15, 12, 15, 3, 15, 15,
+            15, 15, 5, 6, 8, 15, 7, 9, 15, 10, 15, 15, 15, 15, 15, 15,
+            15, 1, 14, 2, 13, 15, 15, 4, 11, 15, 15, 12, 15, 3, 15, 15,
+            15, 15, 5, 6, 8, 15, 7, 9, 15, 10, 15, 15, 15, 15, 15, 15,
+            15, 15, 15, 15, 15, 15, 15, 15, 15, 15, 15, 15, 15, 15, 15, 15,
+            15, 15, 15, 15, 15, 15, 15, 15, 15, 15, 15, 15, 15, 15, 15, 15,
+            15, 15, 15, 15, 15, 15, 15, 15, 15, 15, 15, 15, 15, 15, 15, 15,
+            15, 15, 15, 15, 15, 15, 15, 15, 15, 15, 15, 15, 15, 15, 15, 15,
+            15, 15, 15, 15, 15, 15, 15, 15, 15, 15, 15, 15, 15, 15, 15, 15,
+            15, 15, 15, 15, 15, 15, 15, 15, 15, 15, 15, 15, 15, 15, 15, 15,
+            15, 15, 15, 15, 15, 15, 15, 15, 15, 15, 15, 15, 15, 15, 15, 15,
+            15, 15, 15, 15, 15, 15, 15, 15, 15, 15, 15, 15, 15, 15, 15, 15};
+
+    const size_t len = seq.length();
+
+    // encode bases as nibbles and write them into the alignment struct 2 bases at a time
+    // 4 high bits are first base, 4 low bits are second base
+    uint8_t* bam_seq_start = bam_get_seq(alignment);
+    int i                  = 0;
+    for (i; (i + 1) < len; i += 2)
+    {
+        *bam_seq_start++ = (L[static_cast<unsigned char>(seq[i])] << 4) + L[static_cast<unsigned char>(seq[i + 1])];
+    }
+    // odd number of bases, so we have one left over
+    if (i < len)
+    {
+        *bam_seq_start++ = L[static_cast<unsigned char>(seq[i])] << 4;
+    }
+
+    alignment->core.l_qseq = len;
+
+    return;
+}
+
+static inline void encode_qual(bam1_t* alignment)
+{
+    // if htslib encounters 0xff at the start of the QUAL section, it writes default value (*)
+    auto bam_qual_start = bam_get_qual(alignment);
+    *bam_qual_start     = 0xff;
+}
+
+void print_sam(const std::vector<Overlap>& overlaps,
+               const std::vector<std::string>& cigars,
+               const io::FastaParser& query_parser,
+               const io::FastaParser& target_parser,
+               std::mutex& write_output_mutex,
+               int argc,
+               char* argv[])
+{
+    GW_NVTX_RANGE(profiler, "print_sam::formatting_output");
+
+    // Assume output is stdout, and assume SAM output by default
```

Comment:
how do we choose between sam and bam here?

Reply:
SAM vs BAM depends on adding a parameter to `sam_open()`. Added in latest commit that adds the output enum